### DISCUSSION TITLE
Looking for some Moose classes

### DISCUSSION THREAD

**Discussion by salaudeen-ya** [idaholab/moose#24647] [Q&A General] [answered]
Hello,

Please where can I get the following Moose classes:

`VariableComponent`, `VariableTestComponent`, and `VariablePhiGradient`  with objects `_grad_u`, `_grad_v`, and `_grad_v_phi`  respectively. This is because  I would like to see how the shape functions of their respective objects are computed. If I can see that, I can probably modify it to what I what. In addition, I would be needing that of `MooseArray` too to understand how the object `_q_point` is computed.

I have used things like `git grep VariableGradient`, `git grep VariableTestGradient`, and `git grep VariablePhiGradient` and I still can't find it

**Comment by GiudGiud** [comment]
Hello

I dont think they are input file objects (MooseObjects). They are just part of the code. 

You need to look at the doxygen to understand there things I think.
Here's the link for MooseArray
https://mooseframework.inl.gov/docs/doxygen/moose/classMooseArray.html

You can search for any class there

Guillaume

**Reply by salaudeen-ya** [reply]
I got the following in the `elem.C` file:

```  const FEFamily mapping_family = FEMap::map_fe_type(*this);
  const FEType fe_type(this->default_order(), mapping_family);

  // Build FE and attach quadrature rule.  The default quadrature rule
  // integrates the mass matrix exactly, thus it is overkill to
  // integrate the basis functions, but this is convenient.
  std::unique_ptr<FEBase> fe = FEBase::build(this->dim(), fe_type);
  QGauss qrule (this->dim(), fe_type.default_quadrature_order());
  fe->attach_quadrature_rule(&qrule);

  // Pre-request required data
  const auto & JxW = fe->get_JxW();
  const auto & phi = fe->get_phi();

  // Re-compute element-specific values
  fe->reinit(this);

  // Number of basis functions
  auto N = phi.size();
  libmesh_assert_equal_to(N, this->n_nodes());
```
And the following from the `fe.C` file:

```
  // Most of the hard work happens when we have an actual element
  if (elem)
    {
      // Initialize the shape functions at the user-specified
      // points
      if (pts != nullptr)
        {
          // Set the type and p level for this element
          this->elem_type = elem->type();
          this->_p_level = elem->p_level();

          // Initialize the shape functions
          this->_fe_map->template init_reference_to_physical_map<Dim>
            (*pts, elem);
          this->init_shape_functions (*pts, elem);

          // The shape functions do not correspond to the qrule
          this->shapes_on_quadrature = false;
        }

      // If there are no user specified points, we use the
      // quadrature rule

      // update the type in accordance to the current cell
      // and reinit if the cell type has changed or (as in
      // the case of the hierarchics) the shape functions need

...
```

Now, how do I get to where the shape functions are ACTUALLY computed/implemented?

**Reply by GiudGiud** [reply, marked as answer]
You should keep digging in libmesh.
A lot of the shapes are here I think:
https://mooseframework.inl.gov/docs/doxygen/libmesh/fe__bernstein__shape__1D_8C_source.html#l00193

**Reply by salaudeen-ya** [reply]
So I was trying to dig even more deeper in the libmesh and I found how the Jacobian is multiplied by the weights in the `Assembly.C`, as given below:

```
 // references to the quadrature points and weights
  const std::vector<Real> & JxW = fe->get_JxW();
  const std::vector<Point> & q_points = fe->get_xyz();
```
I tried `git grep get_JxW`, to track where exactly the function is implemented, but all I see is a bunch of pointers

```
framework/src/base/Assembly.C:    (*_holder_fe_helper[dim])->get_JxW();
framework/src/base/Assembly.C:    (*_holder_fe_face_helper[dim])->get_JxW();
framework/src/base/Assembly.C:    (*_holder_fe_face_neighbor_helper[dim])->get_JxW();
framework/src/base/Assembly.C:    (*_holder_fe_neighbor_helper[dim])->get_JxW();
framework/src/base/Assembly.C:    (*_holder_fe_lower_helper[dim])->get_JxW();
framework/src/base/Assembly.C:  _JxW_msm = &_fe_msm->get_JxW();
framework/src/base/Assembly.C:  _current_JxW.shallowCopy(const_cast<std::vector<Real> &>((*_holder_fe_helper[dim])->get_JxW()));
framework/src/base/Assembly.C:      const_cast<std::vector<Real> &>((*_holder_fe_face_helper[dim])->get_JxW()));
framework/src/base/Assembly.C:    const std::vector<Real> & JxW = fe.get_JxW();
...
```

How do I see this `get_JxW` function?

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1FEAbstract.html#afb4de89b31f40d5532ed85da806e2d2e

digging around the fdoxygen is the way to go
